builtin/cat-file.c

```diff
@@ -494,22 +495,42 @@ static int batch_objects(struct batch_options *opt)
 	save_warning = warn_on_object_refname_ambiguity;
 	warn_on_object_refname_ambiguity = 0;
 
-	while (strbuf_getline(&buf, stdin) != EOF) {
-		if (data.split_on_whitespace) {
-			/*
-			 * Split at first whitespace, tying off the beginning
-			 * of the string and saving the remainder (or NULL) in
-			 * data.rest.
-			 */
-			char *p = strpbrk(buf.buf, " \t");
-			if (p) {
-				while (*p && strchr(" \t", *p))
-					*p++ = '\0';
+	if (opt.porcelain_format == PORCELAIN_FORMAT_PORCELAIN) {
+		while (strbuf_getline_nul(&buf, stdin) != EOF) {
+			if (data.split_on_whitespace) {
+				/*
+				* Split at first null, tying off the beginning
+				* of the string and saving the remainder (or NULL) in
+				* data.rest.
+				*/
+				char *p = strpbrk(buf.buf, " \0");
+				if (p) {
+					while (*p && strchr(" \0", *p))
+						*p++ = '\0';
+				}
+				data.rest = p;
 			}
-			data.rest = p;
+
+			batch_one_object(buf.buf, opt, &data);
 		}
+	} else {
+		while (strbuf_getline(&buf, stdin) != EOF) {
+			if (data.split_on_whitespace) {
+				/*
+				* Split at first whitespace, tying off the beginning
+				* of the string and saving the remainder (or NULL) in
+				* data.rest.
+				*/
+				char *p = strpbrk(buf.buf, " \t");
+				if (p) {
+					while (*p && strchr(" \t", *p))
+						*p++ = '\0';
+				}
+				data.rest = p;
+			}
 
-		batch_one_object(buf.buf, opt, &data);
+			batch_one_object(buf.buf, opt, &data);
+		}
```

Comment:
I wonder whether we can somehow do this without changing the indentation, and without duplicating large parts of the code. I have something like this in mind:

```c
struct strbuf buf2 = STRBUF_INIT;
strbuf_getline_fn getline_fn =
    opt.porcelain_format == PORCELAIN_FORMAT_PORCELAIN ?
    strbuf_getline_nul : strbuf_getline;

while (getline_fn(&buf, stdin) != EOF) {
    if (data.split_on_whitespace) {
        /*
         * Split at first whitespace (or NUL), tying off the beginning
         * of the string and saving the remainder (or NUL) in
         * data.rest.
         */
        if (opt.porcelain_format == PORCELAIN_FORMAT_PORCELAIN) {
            if (getline_fn(&buf2, stdin))
                 data.rest = buf2.buf;
        } else {
            char *p = strpbrk(buf.buf, " \t");
            if (p) {
                while (*p && strchr(" \t", *p))
                    *p++ = '\0';
            }
            data.rest = p;
    }
    batch_one_object(buf.buf, opt, &data);
}

strbuf_release(&buf2);
```

What do you think?

Reply:
yeah looks good. I was think about this too and just decided to post something to look at. :smirk: